# Exploring the Ebay Car Sales

We are taking a look at the eBay Kleinanzeigen data set, a classifieds section of the German eBay website. There are sample of 50,000 data points in this dataset. The main goal of this project to clean and perform analysis with this dataset.   

In [2]:
import numpy as np
import pandas as pd 

#Looking at the data types for each column in the dataset
autos = pd.read_csv('autos.csv',encoding='Latin-1')
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [3]:
# Getting a general overview of what kinds of values we can see
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Observations 

After taking a look of the datasets and its data types, we see some insights.

1. There are some columns that contain missing values(gearbox, vehicleType, gearbox, model, fuelType, and notRepairedDamage)
2. There are columns that give out the same value (privat for the seller column, Angebot for the offerType, and 0 for nrOfPictures) based on the first rows
3. The postalCode column is classified as a numerical variable. Usually they would be considered as text, because they do not provide any signficance in performing calculations

## Cleaning Up the Column Names

We want to make changes to the formatting of the column headers from camelcase to snakecase (the preferred way in Python). Plus, it makes it more readable.

In [4]:
#Creating an new array with the edited column headers
autos_copy = autos.copy()
new_columns = []

def cleaning(string):
    
    #Manually replacing some of the column headers
    string = string.replace("powerPS", "power_ps")
    string = string.replace("yearOfRegistration","registration_year")
    string = string.replace("monthOfRegistration", "registration_month")
    string = string.replace("notRepairedDamage", "unrepaired_damage")
    string = string.replace("dateCreated", "date_created")
    
    #Using a list comprehension for the camelcase to snakecase conversion
    #for the rest of the column headers. List comprehension is more compact
    #of creating a new list rather than using an entire for loop.
    string = ''.join(['_'+i.lower() if i.isupper()  
               else i for i in string]).lstrip('_')  
    
    return string

for columns in autos_copy:
    clean_columns = cleaning(columns)
    new_columns.append(clean_columns) 

autos.columns = new_columns
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

## **Looking More into the Dataset**

In [5]:
autos['seller'].head

<bound method Series.head of 0        privat
1        privat
2        privat
3        privat
4        privat
5        privat
6        privat
7        privat
8        privat
9        privat
10       privat
11       privat
12       privat
13       privat
14       privat
15       privat
16       privat
17       privat
18       privat
19       privat
20       privat
21       privat
22       privat
23       privat
24       privat
25       privat
26       privat
27       privat
28       privat
29       privat
          ...  
49970    privat
49971    privat
49972    privat
49973    privat
49974    privat
49975    privat
49976    privat
49977    privat
49978    privat
49979    privat
49980    privat
49981    privat
49982    privat
49983    privat
49984    privat
49985    privat
49986    privat
49987    privat
49988    privat
49989    privat
49990    privat
49991    privat
49992    privat
49993    privat
49994    privat
49995    privat
49996    privat
49997    privat
49998    privat
49999    pr

In [6]:
autos['offer_type'].head

<bound method Series.head of 0        Angebot
1        Angebot
2        Angebot
3        Angebot
4        Angebot
5        Angebot
6        Angebot
7        Angebot
8        Angebot
9        Angebot
10       Angebot
11       Angebot
12       Angebot
13       Angebot
14       Angebot
15       Angebot
16       Angebot
17       Angebot
18       Angebot
19       Angebot
20       Angebot
21       Angebot
22       Angebot
23       Angebot
24       Angebot
25       Angebot
26       Angebot
27       Angebot
28       Angebot
29       Angebot
          ...   
49970    Angebot
49971    Angebot
49972    Angebot
49973    Angebot
49974    Angebot
49975    Angebot
49976    Angebot
49977    Angebot
49978    Angebot
49979    Angebot
49980    Angebot
49981    Angebot
49982    Angebot
49983    Angebot
49984    Angebot
49985    Angebot
49986    Angebot
49987    Angebot
49988    Angebot
49989    Angebot
49990    Angebot
49991    Angebot
49992    Angebot
49993    Angebot
49994    Angebot
49995    Angebot
49

In [7]:
autos['nr_of_pictures'].head

<bound method Series.head of 0        0
1        0
2        0
3        0
4        0
5        0
6        0
7        0
8        0
9        0
10       0
11       0
12       0
13       0
14       0
15       0
16       0
17       0
18       0
19       0
20       0
21       0
22       0
23       0
24       0
25       0
26       0
27       0
28       0
29       0
        ..
49970    0
49971    0
49972    0
49973    0
49974    0
49975    0
49976    0
49977    0
49978    0
49979    0
49980    0
49981    0
49982    0
49983    0
49984    0
49985    0
49986    0
49987    0
49988    0
49989    0
49990    0
49991    0
49992    0
49993    0
49994    0
49995    0
49996    0
49997    0
49998    0
49999    0
Name: nr_of_pictures, Length: 50000, dtype: int64>

Looking back at the initial observations, we said that that offer_type, seller, and nr_of_pictures only contained one value justed based on the first few datapoints. Now, after looking at those columns for the entire dataset, each did have one particular value. We can delete those columns out. 

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

               date_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [9]:
#Checking any discrepenices in the model names
autos['model'].unique()

array(['andere', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'fiesta',
       'wrangler', 'clubman', 'a6', 'transporter', 'astra', 'v40',
       'ibiza', 'micra', '1er', 'yaris', 'colt', '6_reihe', '5_reihe',
       'corolla', 'ka', 'tigra', 'punto', 'vito', 'cordoba', 'galaxy',
       '100', 'sharan', 'octavia', 'm_klasse', 'lupo', 'superb', 'meriva',
       'c_max', 'laguna', 'touran', '1_reihe', 'm_reihe', 'touareg',
       'seicento', 'avensis', 'vivaro', 'x_reihe', 'ducato', 'carnival',
       'boxster', 'signum', 'zafira', 'rav', 'a5', 'beetle', 'c_reihe',
       'phaeton', 'i_reihe', 'sl', 'insignia', 'up', 'civic', '80',
       'mx_reihe', 'omega', 'sorento', 'z_reihe', 'berlingo', 'clk',
       '

In [10]:
#Wanted to look what car models were considered kleinwagen 
#(don't know German)
autos['model'].loc[autos['vehicle_type'] == 'kleinwagen']

3         fortwo
12        fortwo
14          clio
26          polo
28        cooper
30       2_reihe
32         corsa
42          polo
43            a1
44        twingo
45        andere
56          clio
63        cooper
66         corsa
69       sandero
78       3_reihe
79         ibiza
81         micra
85         yaris
102           ka
103        micra
107        punto
112         polo
116        corsa
127         polo
128         lupo
130       fiesta
146       fiesta
148         polo
156      1_reihe
          ...   
49866       golf
49873       polo
49874         ka
49886     almera
49889      ibiza
49891         a2
49895      arosa
49901        one
49905       lupo
49908      panda
49915     twingo
49920     andere
49923      yaris
49924      ibiza
49929      corsa
49940        3er
49949       golf
49951      corsa
49952       polo
49956      yaris
49960         ka
49966         c1
49971       lupo
49975       jazz
49979       polo
49982      fabia
49983      focus
49989       po

In [11]:
#Could be converted into a boolean data type
autos['unrepaired_damage'].head

<bound method Series.head of 0        nein
1        nein
2        nein
3        nein
4        nein
5         NaN
6         NaN
7        nein
8        nein
9        nein
10       nein
11        NaN
12       nein
13       nein
14        NaN
15       nein
16       nein
17       nein
18        NaN
19       nein
20        NaN
21       nein
22       nein
23       nein
24       nein
25         ja
26       nein
27        NaN
28         ja
29       nein
         ... 
49970    nein
49971    nein
49972      ja
49973    nein
49974    nein
49975    nein
49976    nein
49977    nein
49978      ja
49979    nein
49980     NaN
49981    nein
49982     NaN
49983     NaN
49984     NaN
49985     NaN
49986    nein
49987    nein
49988    nein
49989      ja
49990    nein
49991     NaN
49992    nein
49993     NaN
49994    nein
49995    nein
49996    nein
49997    nein
49998    nein
49999    nein
Name: unrepaired_damage, Length: 50000, dtype: object>

In [12]:
autos['price'].head

<bound method Series.head of 0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
5         $7,900
6           $300
7         $1,990
8           $250
9           $590
10          $999
11          $350
12        $5,299
13        $1,350
14        $3,999
15       $18,900
16          $350
17        $5,500
18          $300
19        $4,150
20        $3,500
21       $41,500
22       $25,450
23        $7,999
24       $48,500
25           $90
26          $777
27            $0
28        $5,250
29        $4,999
          ...   
49970    $15,800
49971       $950
49972     $3,300
49973     $6,000
49974         $0
49975     $9,700
49976     $5,900
49977     $5,500
49978       $900
49979    $11,000
49980       $400
49981     $2,000
49982     $1,950
49983       $600
49984         $0
49985     $1,000
49986    $15,900
49987    $21,990
49988     $9,550
49989       $150
49990    $17,500
49991       $500
49992     $4,800
49993     $1,650
49994     $5,000
49995    $24,900
49

Other things to take note of:
1. The price and odometer columns need to be converted into numerical data
2. The max value for registration_year is 9999, which needs to be looked at
3. We can convert the unrepaired_damage column into a boolean data type (since it asks for a yes or no)
4. There are some prices that have been listed as $0 (look into after the conversion)

**Removing out Unnecessary Columns**

In [13]:
#Now we have a total of 17 columns
autos = autos.drop(['seller', 'offer_type','nr_of_pictures'],axis=1)
autos.shape

(50000, 17)

**Column Conversions**

In [14]:
#Removing the unnecessary characters 
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(",",'')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')

#Converting the string columns into int
autos['price'] = autos['price'].astype(int)
autos['odometer'] = autos['odometer'].astype(int)

#Making sure they are the correct type
autos['price'].dtype
autos['odometer'].dtype

#Converting the unrepaired_damage column to a boolean
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('nein','0')
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('ja','1')
autos['unrepaired_damage'] = autos['unrepaired_damage'].astype(bool)
autos['unrepaired_damage'].dtype

#Converting the postal_code code to string
autos['postal_code'] = autos['postal_code'].astype(str)

#renaming the odometer column 
autos.rename(columns={"odometer": "odometer_km"})



date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

## Taking a Look at the Price and Odometer Columns

In [15]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos['price'].unique().shape

(2357,)

In [18]:
#Taking a look at the counts of car prices 
autos['price'].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

We see here that there some high end prices, mainly the 7 digit car prices, are the outliers. The reason here is because not many people can't pay for those kinds of prices, unless your a rich car collector. Not only that, it is often rare to see these kinds of car prices.  

In [20]:
#Deleting out the outliers and checking how many datapoints 
#are there without those outliers
autos = autos[autos["price"].between(0,3890000)]
autos.shape

(49992, 17)

In [21]:
#Looking at the summary stats of the odometer column
autos['odometer'].describe()

count     49992.000000
mean     125734.017443
std       40041.246220
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64

In [22]:
autos['odometer'].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       818
50000      1026
60000      1164
70000      1230
80000      1436
90000      1757
100000     2168
125000     5170
150000    32419
Name: odometer, dtype: int64

The outlier in this column would be cars that have 150,000km on the odometer. That is because there have almost 6 times as many cars compared to second highest count, which have 125,000km on the odometer.  

In [23]:
#Deleting out the outlier for this
autos = autos[autos["odometer"].between(0,125000)]
autos.shape

(17573, 17)

## Looking at the Date Columns

In [25]:
#Getting an overview of how the data is formatted
autos[['date_crawled','date_created','last_seen']][0:5]

date_crawled         date_created            last_seen
2   2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3   2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
12  2016-03-31 19:48:22  2016-03-31 00:00:00  2016-04-06 14:17:52
15  2016-04-01 12:06:20  2016-04-01 00:00:00  2016-04-02 21:10:48
22  2016-03-28 20:50:54  2016-03-28 00:00:00  2016-04-01 06:45:30

**Looking at the date_crawled column**

In [31]:
date_crawled = autos['date_crawled'].str[:10]
date_crawled

2        2016-03-26
3        2016-03-12
12       2016-03-31
15       2016-04-01
22       2016-03-28
24       2016-04-03
26       2016-04-03
37       2016-03-23
38       2016-03-21
42       2016-03-22
43       2016-03-08
49       2016-03-27
50       2016-03-09
52       2016-03-25
53       2016-04-02
54       2016-03-15
55       2016-03-07
56       2016-03-18
58       2016-03-30
62       2016-03-12
63       2016-03-14
65       2016-04-04
68       2016-04-03
69       2016-03-21
71       2016-03-28
73       2016-03-18
76       2016-03-22
85       2016-03-11
86       2016-04-03
89       2016-04-05
            ...    
49918    2016-04-03
49921    2016-03-05
49923    2016-03-05
49924    2016-03-21
49929    2016-03-27
49931    2016-03-22
49933    2016-04-03
49934    2016-03-11
49939    2016-03-13
49941    2016-03-15
49942    2016-03-09
49948    2016-03-22
49950    2016-03-15
49952    2016-03-21
49953    2016-03-30
49956    2016-03-28
49963    2016-03-26
49965    2016-03-11
49970    2016-03-21


In [33]:
#Calculating the the distribution of values
date_crawled.value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.028567
2016-03-06    0.013259
2016-03-07    0.034599
2016-03-08    0.030957
2016-03-09    0.031184
2016-03-10    0.032607
2016-03-11    0.033119
2016-03-12    0.038070
2016-03-13    0.016446
2016-03-14    0.036590
2016-03-15    0.032152
2016-03-16    0.029079
2016-03-17    0.029932
2016-03-18    0.013145
2016-03-19    0.038696
2016-03-20    0.038127
2016-03-21    0.035281
2016-03-22    0.032607
2016-03-23    0.030729
2016-03-24    0.029989
2016-03-25    0.030957
2016-03-26    0.032948
2016-03-27    0.031810
2016-03-28    0.035224
2016-03-29    0.036021
2016-03-30    0.032891
2016-03-31    0.031696
2016-04-01    0.035850
2016-04-02    0.036420
2016-04-03    0.040915
2016-04-04    0.033688
2016-04-05    0.012178
2016-04-06    0.002390
2016-04-07    0.001878
Name: date_crawled, dtype: float64

The web crawler got most of the datapoints on April 3rd, 2016 and got the least amount on April 7th, 2016, which is very intresting since they are only a few days apart. 

**Looking at the date_created column**

In [34]:
date_created = autos['date_created'].str[:10]
date_created

2        2016-03-26
3        2016-03-12
12       2016-03-31
15       2016-04-01
22       2016-03-28
24       2016-04-03
26       2016-04-03
37       2016-03-23
38       2016-03-21
42       2016-03-22
43       2016-03-08
49       2016-03-27
50       2016-03-09
52       2016-03-25
53       2016-04-02
54       2016-03-15
55       2016-03-06
56       2016-03-18
58       2016-03-30
62       2016-03-12
63       2016-03-14
65       2016-04-04
68       2016-04-03
69       2016-03-21
71       2016-03-28
73       2016-03-18
76       2016-03-22
85       2016-03-11
86       2016-04-03
89       2016-04-05
            ...    
49918    2016-04-03
49921    2016-02-16
49923    2016-03-05
49924    2016-03-21
49929    2016-03-27
49931    2016-03-22
49933    2016-04-03
49934    2016-03-11
49939    2016-03-13
49941    2016-03-14
49942    2016-03-09
49948    2016-03-22
49950    2016-03-15
49952    2016-03-21
49953    2016-03-30
49956    2016-03-28
49963    2016-03-26
49965    2016-03-11
49970    2016-03-21


In [35]:
date_created.value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2015-06-11    0.000057
2015-08-10    0.000057
2015-09-09    0.000057
2015-11-10    0.000057
2015-12-30    0.000057
2016-01-03    0.000057
2016-01-10    0.000057
2016-01-22    0.000057
2016-01-27    0.000114
2016-02-01    0.000057
2016-02-05    0.000057
2016-02-07    0.000057
2016-02-09    0.000057
2016-02-14    0.000057
2016-02-16    0.000057
2016-02-17    0.000057
2016-02-18    0.000057
2016-02-19    0.000114
2016-02-20    0.000057
2016-02-21    0.000114
2016-02-23    0.000228
2016-02-24    0.000057
2016-02-25    0.000171
2016-02-26    0.000057
2016-02-27    0.000114
2016-02-28    0.000341
2016-02-29    0.000228
2016-03-01    0.000057
2016-03-02    0.000171
2016-03-03    0.000797
                ...   
2016-03-09    0.031810
2016-03-10    0.032038
2016-03-11    0.033290
2016-03-12    0.037842
2016-03-13    0.017755
2016-03-14    0.035338
2016-03-15    0.032209
2016-03-16    0.029534
2016-03-17    0.029534
2016-03-18    0.013828
2016-03-19    0.037159
2016-03-20    0.038297
2016-03-21 

We see that from June 16th of 2015 till the beginning days of March in 2016, there were not alot of cars listed for sale, with lowest percentage being on Feburary 20th as well as March 1st and highest being on March 3rd. However, there is an upward of car listings from the from March 9th, 2016 till April 7th,2016, with the highest percentange coming on April 3rd and the lowest on April 7th, which is quite interesting, since they are only a few days apart. 

**Looking at the last_seen column**

In [37]:
last_seen = autos['last_seen'].str[:10]
last_seen

2        2016-04-06
3        2016-03-15
12       2016-04-06
15       2016-04-02
22       2016-04-01
24       2016-04-07
26       2016-04-05
37       2016-04-05
38       2016-03-28
42       2016-04-06
43       2016-04-07
49       2016-04-05
50       2016-04-07
52       2016-04-07
53       2016-04-06
54       2016-03-30
55       2016-03-08
56       2016-03-31
58       2016-04-01
62       2016-04-03
63       2016-04-06
65       2016-04-05
68       2016-04-05
69       2016-03-28
71       2016-04-07
73       2016-04-05
76       2016-04-06
85       2016-04-07
86       2016-04-03
89       2016-04-07
            ...    
49918    2016-04-05
49921    2016-04-05
49923    2016-04-07
49924    2016-03-24
49929    2016-04-07
49931    2016-04-06
49933    2016-04-06
49934    2016-04-06
49939    2016-04-06
49941    2016-03-30
49942    2016-04-05
49948    2016-04-06
49950    2016-03-21
49952    2016-04-07
49953    2016-03-30
49956    2016-03-30
49963    2016-03-28
49965    2016-03-15
49970    2016-04-07


In [38]:
last_seen.value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.001024
2016-03-06    0.003756
2016-03-07    0.004097
2016-03-08    0.005235
2016-03-09    0.007568
2016-03-10    0.009674
2016-03-11    0.009617
2016-03-12    0.021567
2016-03-13    0.007113
2016-03-14    0.010414
2016-03-15    0.014568
2016-03-16    0.013088
2016-03-17    0.023673
2016-03-18    0.007568
2016-03-19    0.013487
2016-03-20    0.018153
2016-03-21    0.017982
2016-03-22    0.019120
2016-03-23    0.016787
2016-03-24    0.017072
2016-03-25    0.017356
2016-03-26    0.015194
2016-03-27    0.014226
2016-03-28    0.018494
2016-03-29    0.021567
2016-03-30    0.022876
2016-03-31    0.021681
2016-04-01    0.022022
2016-04-02    0.024014
2016-04-03    0.024299
2016-04-04    0.022136
2016-04-05    0.137882
2016-04-06    0.246799
2016-04-07    0.149889
Name: last_seen, dtype: float64

Looking at the distribution, we see that the web crawler has seen more of the ads recently was between April 5th to the 7th of 2016. On the flip side, the web crawler only has seen ads recently between the days of March 5th to March 11th of 2016.

**Correcting the registration year column**

In [39]:
autos['registration_year'].describe()

count    17573.000000
mean      2010.057930
std        169.977074
min       1000.000000
25%       2002.000000
50%       2008.000000
75%       2011.000000
max       9999.000000
Name: registration_year, dtype: float64

We see that min and max values for the registration year don't really quite make any sense for the dataset. Let's take a look the number of cars between those values  

In [43]:
autos['registration_year'].value_counts(normalize=True).sort_index(ascending=True)

1000    0.000057
1001    0.000057
1111    0.000057
1500    0.000057
1800    0.000114
1910    0.000455
1927    0.000057
1929    0.000057
1931    0.000057
1934    0.000114
1937    0.000171
1938    0.000057
1939    0.000057
1941    0.000114
1943    0.000057
1948    0.000057
1950    0.000171
1951    0.000114
1952    0.000057
1954    0.000114
1955    0.000114
1956    0.000285
1957    0.000114
1958    0.000171
1959    0.000285
1960    0.001536
1961    0.000228
1962    0.000171
1963    0.000228
1964    0.000683
          ...   
2001    0.029762
2002    0.031127
2003    0.036590
2004    0.037899
2005    0.049394
2006    0.053662
2007    0.053207
2008    0.066010
2009    0.079554
2010    0.066409
2011    0.075229
2012    0.067604
2013    0.043305
2014    0.037615
2015    0.022535
2016    0.018039
2017    0.022933
2018    0.008536
2019    0.000057
2800    0.000057
4100    0.000057
4500    0.000057
4800    0.000057
5000    0.000228
5911    0.000057
6200    0.000057
8888    0.000057
9000    0.0000

Taking a look deeper, we can see that there cars that have been registered beyond 2019 and before 1800. Based off of that, let's have the lowest value be 1800 and the highest be 2019, since that is range when cars are around. 

In [45]:
#Removing out the years before 1800 and after 2019
autos = autos[autos["registration_year"].between(1800,2019)]
autos["registration_year"].describe()

count    17552.000000
mean      2005.580276
std          9.226827
min       1800.000000
25%       2002.000000
50%       2008.000000
75%       2011.000000
max       2019.000000
Name: registration_year, dtype: float64